source/particle/generator/interface.cc

```diff
@@ -0,0 +1,191 @@
+/*
+  Copyright (C) 2015 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/interface.h>
+
+#include <deal.II/base/std_cxx1x/tuple.h>
+#include <deal.II/grid/grid_tools.h>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      std::pair<types::LevelInd,Particle<dim> >
+      Interface<dim>::generate_particle(const Point<dim> &position,
+                                        const types::particle_index id) const
+      {
+        const typename parallel::distributed::Triangulation<dim>::active_cell_iterator it =
+          (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), position)).first;;
+
+        //Only try to add the point if the cell it is in, is on this processor
+        if (it->is_locally_owned())
+          {
+            const Particle<dim> particle(position, id);
+            const types::LevelInd cell(it->level(), it->index());
+            return std::make_pair(cell,particle);
+          }
+
+        AssertThrow(false,
+                    ExcMessage("A particle generator plugin tried to create a particle in"
```

Comment:
By the way, this seems like an exception that is specific enough that one could also declare an exception class for it, using
```
  DeclExceptionMsg (ExcNonlocalPoint, ....)
```
and then throw the exception by name, rather than using a generic `ExcMessage`. This would also make it simpler and more descriptive to catch just this specific exception in all of the places where that's necessary.